notesapi/v1/views.py

```diff
@@ -114,82 +118,136 @@ class AnnotationSearchView(GenericAPIView):
+    @property
+    def paginator(self):
+        """
+        The paginator instance associated with the view and used data source, or `None`.
+        """
+        if not hasattr(self, '_paginator'):
+            if self.pagination_class is None:
+                self._paginator = None
+            else:
+                self._paginator = self.pagination_class() if self.is_es_disabled else ESNotesPagination()
+
+        return self._paginator
 
-    def get_from_db(self, *args, **kwargs):  # pylint: disable=unused-argument
+    def filter_queryset(self, queryset):
         """
-        Search annotations in database.
+        Given a queryset, filter it with whichever filter backend is in use.
+
+        Do not filter additionally if mysql db used or use `CompoundSearchFilterBackend`
+        and `HighlightBackend` if elasticsearch is the data source.
         """
-        query = Note.objects.filter(**self.query_params).order_by('-updated')
+        filter_backends = []
+        if not self.is_es_disabled:
+            filter_backends = [CompoundSearchFilterBackend]
+            if self.params.get('highlight'):
+                filter_backends.append(HighlightBackend)
 
-        if 'text' in self.params:
-            query = query.filter(Q(text__icontains=self.params['text']) | Q(tags__icontains=self.params['text']))
+        for backend in list(filter_backends):
+            queryset = backend().filter_queryset(self.request, queryset, view=self)
+        return queryset
 
+    def list(self, *args, **kwargs):
+        """
+        Returns list of students notes.
+        """
         # Do not send paginated result if usage id based search.
         if self.search_with_usage_id:
-            serializer = NoteSerializer(query, many=True)
+            queryset = self.filter_queryset(self.get_queryset())
+            serializer = self.get_serializer(queryset, many=True)
             return Response(serializer.data, status=status.HTTP_200_OK)
+        return super(AnnotationSearchView, self).list(*args, **kwargs)
 
-        page = self.paginate_queryset(query)
-        serializer = NoteSerializer(page, many=True)
-        response = self.get_paginated_response(serializer.data)
-        return response
-
-    def get_from_es(self, *args, **kwargs):  # pylint: disable=unused-argument
+    def get(self, *args, **kwargs):  # pylint: disable=unused-argument
         """
-        Search annotations in ElasticSearch.
+        Search annotations in most appropriate storage
         """
-        query = SearchQuerySet().models(Note).filter(**self.query_params)
+        self.query_params = {}
+        self.search_with_usage_id = False
+        self.params = self.request.query_params.dict()
+        usage_ids = self.request.query_params.getlist('usage_id')
 
-        if 'text' in self.params:
-            clean_text = query.query.clean(self.params['text'])
-            query = query.filter(SQ(data=clean_text))
+        if len(usage_ids):
```

Comment:
I would check boolean of the usage_ids `len` operation looks redundant. What do you think?

Reply:
Here the situation is obscure and I'm not 100% sure that we can do it like `if usage_ids:`. I thought about it, I didn’t do that. This condition was already in the old implementation and I had to port it to a new one. This is part of the business logic that I do not know to the all details.
It used to be even like this:
```
if len(usage_ids) > 0:
```
I simplified it safely as much as I had the courage. Here, as I understand it, it is necessary to check that usage_ids is a sequence and that the sequence is not empty.
To simplify to `if usage_ids`, this will not be a sequence check. With usage_ids = 12, this condition will work, which is probably not true for logic. Though `len('foo')` will work out and return True, which is also a concern. Let's leave it as it is because the logic `self.query_params ['usage_id__in'] = usage_ids` below is afraid of breaking it.

Reply:
usage_ids is a result of `getlist` which guarantees to return a list or an empty list if there is no default value.